messagecatalog/spec.md

```diff
@@ -0,0 +1,1475 @@
+stored and shared in a single document. A folder in a git repository containing
+such documents is also a catalog in the sense of this specification.
+
+For interoperability and sharing design-time definitions across system
+boundaries, conformance to a common data model and document format is most
+important and it is assumes that the authoritative source for many definitions
+will into be a document under source control that is then shared at runtime via
+the common API defined here.
+
+## 2. Message Catalog Elements
+
+This section describes the elements enumerated in the introduction.
+
+All data types used in this section are defined in the CNCF CloudEvents
+specification and MUST follow the respective formatting and syntax rules unless
+specified otherwise.
+
+These type constraints permit for the message catalog definitions to be
+consistently mapped into the data encodings supported by CloudEvents,
+specifically also into Apache Avro, JSON, Protobuf, and AMQP. All examples in
+this document use JSON notation.
+
+### 2.1. Message definition group
+
+Since message definitions are often used across several applications and in
+conjunction with related message definitions like commands and command replies,
+and those definitions might even be shared by multiple organizations, it is very
+useful to put a grouping construct around sets of message definitions that are
+related either by ownership or by a shared subject matter context.
+
+Implementations of this specification MAY associate access control rules with
+message definition groups. For instance, a user or a group of users might be
+given write access only to a particular message definition group that their
+organization owns. If trade secret protection is necessary for an application or
+parts of it and the definition structure would give some of those away, read
+access to a group of definitions might likewise be restricted.
+
+This specification does not define management constructs for access control
+rules.
+
+### 2.1.1. Message definition group attributes
+
+The data model for a message definition group consists of these attributes:
+
+#### `id` (message definition group)
+
+- Type: `String`
+- Description: Identifies the message definition group.
+- Constraints:
+  - REQUIRED. IMMUTABLE.
+  - MUST be a non-empty string
+  - MUST conform with RFC3986/3.3 `segment-nz-nc` syntax. This allows for "dot
+    notation", e.g. `org.example.myapp.module` for logical organization of
+    message definition groups, even though the namespace is flat.
+  - MUST be unique within the scope of the message catalog
+- Examples:
+  - mygroup
+  - my-group
+  - org.example.myapp.module
+  - events\@example.com
+
+#### `specversion` (message definition group)
+
+- Type: `String`
+- Description: Version of this specification that this document complies with.
+- Constraints:
+  - REQUIRED.
+  - MUST be "0.1" for this version of the specification.
+
+#### `format` (message definition group)
```

Comment:
to support several formats for the same group set, I would have to make several message groups?